### PR TITLE
 Introduced auto. spam learning & global sieve filter

### LINKED ISSUE #384: Add support for dovecot-antispam plugin
http://wiki2.dovecot.org/Plugins/Antispam

> The antispam plugin allows you to retrain the spam filter by simply moving emails in and out of the Spam folder.

Can this be added to the container?  Running the `sa-learn` process by hand or via cron is a little clumsy.

### PR DESCRIPTION
**Pull request by 17Halbe**
#### Introduced auto-spam-learning via sieve. 
Spam is automatically trained to spamassassin, when moved via **imap-only** to the spam folder.
And so is "Ham" when it's moved out of the Spam folder.

Additionally two new optional config files have been introduced:
- global-before.dovecot.sieve
- global-after.dovecot.sieve

The former being a global sieve file which is going to be executed before the user specific rules are applied and the later, you might have guessed it, is applied, after the user specific rules have been executed.

**Attention:**

This has not been tested with actual Spam, Ham movement. We need a testcase or at least someone to test it before merging, alone I don't know how to programmatically move mails between two imap-folders. Anyone else has a hint? Or can try this out?

Tests are passing on my machine.
closes  #384

### CODE REVIEW

File: target/dovecot/90-sieve.conf [hunk 1]
```diff
@@ -108,4 +108,18 @@ plugin {
   # Locations of programs that can be called by the sieve_extprograms plugin
   sieve_pipe_bin_dir = /usr/lib/dovecot/sieve-pipe
   sieve_filter_bin_dir = /usr/lib/dovecot/sieve-filter
+
+  # AntispamWithSieve (Derived from: https://wiki2.dovecot.org/HowTo/AntispamWithSieve)
+
+  # From elsewhere to Spam folder
+  imapsieve_mailbox1_name = Spam
```
Comment (johansmitsnl) [inline target/dovecot/90-sieve.conf]: Should this be Junk?
https://github.com/tomav/docker-mailserver/blob/a7670ac5c122e3ed126dc263b9d23ae3b2997a79/test/config/dovecot-lmtp/conf.d/15-mailboxes.conf#L23

File: target/dovecot/90-sieve.conf [hunk 2]
```diff
@@ -108,4 +108,18 @@ plugin {
...
+  imapsieve_mailbox1_causes = COPY
+  imapsieve_mailbox1_before = file:/usr/lib/dovecot/antispam/report-spam.sieve
+
+  # From Spam folder to elsewhere
+  imapsieve_mailbox2_name = *
+  imapsieve_mailbox2_from = Spam
```
Comment (johansmitsnl) [inline target/dovecot/90-sieve.conf]: Should this be Junk?
https://github.com/tomav/docker-mailserver/blob/a7670ac5c122e3ed126dc263b9d23ae3b2997a79/test/config/dovecot-lmtp/conf.d/15-mailboxes.conf#L23

File: target/dovecot/antispam/.sa-learn-ham.sh [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+exec /usr/bin/sa-learn -u ${1} --ham
```
Comment (johansmitsnl) [inline target/dovecot/antispam/.sa-learn-ham.sh]: Add a newline to the end of the file

File: target/dovecot/antispam/.sa-learn-spam.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+exec /usr/bin/sa-learn -u ${1} --spam
```
Comment (johansmitsnl) [inline target/dovecot/antispam/.sa-learn-spam.sh]: Add a newline to the end of the file

File: target/dovecot/antispam/report-ham.sieve [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+if string "${mailbox}" "Trash" {
+  stop;
+}
+
+if environment :matches "imap.user" "*" {
+  set "username" "${1}";
+}
+
+pipe :copy ".sa-learn-ham.sh" [ "${username}" ];
```
Comment (johansmitsnl) [inline target/dovecot/antispam/report-ham.sieve]: Add a newline to the end of the file

File: target/dovecot/antispam/report-spam.sieve [hunk 6]
```diff
@@ -0,0 +1,7 @@
+require ["vnd.dovecot.pipe", "copy", "imapsieve", "environment", "variables"];
+
+if environment :matches "imap.user" "*" {
+  set "username" "${1}";
+}
+
+pipe :copy ".sa-learn-spam.sh" [ "${username}" ];
```
Comment (johansmitsnl) [inline target/dovecot/antispam/report-spam.sieve]: Add a newline to the end of the file

**Comment by johansmitsnl** [comment]
Good PR and much wanted!

Not sure how to test this though.

File: target/dovecot/90-sieve.conf (hunk 1 above)
Comment (17Halbe) [reply target/dovecot/90-sieve.conf]: You are so right.. Never base decisions on assumptions!!

**Comment by 17Halbe** [comment]
I don't know on those tests. There is definitely a way to test that programmatically, client mail programs do that all the time, I'll be having a lot less sparetime from tomorrow on until I guess mid September! O_o 
So I won't be able to dive into "programming a imap-compatible mail client" ;) 
But my suggestion would be, to just merge it and give it a try. It's not the nicest solution, but it might even work as intended!

**Comment by johansmitsnl** [comment]
I tried testing but I have 2 questions:
1. Should the dbpath be set to: `--dbpath /var/mail-state/lib-amavis/.spamassassin`  for the sa-learn [wiki](https://github.com/tomav/docker-mailserver/wiki/FAQ-and-Tips#how-can-i-make-spamassassin-learn-spam)?
2. Unknown what the argument of the learn scripts are, but you should pass a file or a directory which you want to train or is this done trough the pipe?

**Comment by 17Halbe** [comment]
Yes, the respective mail will be piped to sa-learn.

And you are right. dbpath has to be set, I wasn't actually aware that we weren't using user specific spam-DBs.
Does that make sense though?
It's probably something to discuss somewhere else. But it might be worth considering giving every user it's own spam db?! (I'm ambivalent about it)
So for this PR I'll of course add the dbpath option. :+1: 
(So I can omit the user override as well.. :'( )
It just came to my mind, that we as well might have a permission problem... (Dovecot is executing sa-learn.) 
I'll have a look the day after tomorrow..

**Comment by 17Halbe** [comment]
Please don't merge yet. I'll give this a try on a real mailserver..

**Comment by 17Halbe** [comment]
Since learning every single mail at the time of movement is quite expensive (SA has to be started for every mail) I would suggest moving spam mails to a separate folder and delay learning to once a day by cron. 
Any objections? Should we pick a specific time during the night or just a random @daily?

**Comment by johansmitsnl** [comment]
That sounds a good idea. On the wiki there is already something about this but it is only for one folder, not for all folders.

**Comment by johansmitsnl** [comment]
@17Halbe will you update this branch or do you provide a new PR?

**Comment by 17Halbe** [comment]
Sorry for leaving you hanging! I really only get to work on this project while I‘m travelling. There‘s a lot of work related stuff I got to do till Friday. Thereafter I‘ll probably got a week off. 
I‘ll update this branch. It‘s not been forgotten! ;)

**Comment by johansmitsnl** [comment]
@17Halbe was just asking, no hurries ;)

**Comment by 17Halbe** [comment]
Sorry for the "long time no see". I was sick and away for it. I won't be able to contribute for some time. Sorry. Feel free to close this one.

**Pull request closed without merging**